src/com/ka/spreadsheet/diff/UnifiedDiffSpreadSheetDiffCallback.java

```diff
@@ -0,0 +1,172 @@
+package com.ka.spreadsheet.diff;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.ka.spreadsheet.diff.SpreadSheetUtils.CELL_INTERNAL_TO_USER;
+
+// Output format, inspired by the traditional Unified Diff format (but using cell references instead of line numbers):
+//      +--- column 1
+//      V
+//        --- file_1_name!sheet_1_name
+//        +++ file_2_name!sheet_1_name
+//        @@ cell_1_ref,cell_n_ref cell_1_ref,cell_n_ref @@
+//        -old_cell_1_data
+//        -...
+//        -old_cell_n_data
+//        +new_cell_1_data
+//        +...
+//        +new_cell_n_data
+//        @@@ ...
+//        ...
+//        --- file_1_name!sheet_2_name
+//        +++ file_2_name!sheet_2_name
+//        ...
+// Each cell data line is always present, even if the data is empty (thus just "-" or "+).
+public class UnifiedDiffSpreadSheetDiffCallback implements SpreadSheetDiffCallback {
+
+  private final String lineSeparator = System.getProperty("line.separator");
+  private List<DiffCell> diffCells = new ArrayList<DiffCell>();
+
+  // This formatter requires that each cell in the pair of workbooks will be processed in worksheet-row-column order.
+  // SpreadSheetDiffer.java does that, but it's not a contract.  If that changes some day, set this to true to resort
+  // the list of diffs/extras before formatting them.
+  private static final boolean diffCellsIsUnsorted = false;
+
+  @Override
+  public void reportWorkbooksDiffer(boolean differ, File file1, File file2) {
+    String sheetName = "";
+    int row = -1;
+    int col = -1;
+    List<DiffCell> cellBlock = null;
+    sortList(diffCells);
+    for (DiffCell diffCell : diffCells) {
+     if (!diffCell.sheetName.equals(sheetName)) {
+        // New sheet, finish the active row (if any) and start the new sheet.
+        if (cellBlock != null) {
+          printRow(cellBlock);
+        }
+        sheetName = diffCell.sheetName;
+        System.out.println("--- " + file1 + "!" + sheetName);
+        System.out.println("+++ " + file2 + "!" + sheetName);
+        cellBlock = null;
+        row = -1;
+      }
+      if (diffCell.rowIndex != row) {
+        // New row, finish the active row (if any) and start the new row.
+        if (cellBlock != null) {
+          printRow(cellBlock);
+        }
+        cellBlock = null;
+        row = diffCell.rowIndex;
+        col = -1;
+      }
+      if (diffCell.colIndex != col) {
+        // New cell-block, finish the active cell-block (if any) and start a new cell-block.
+        if (cellBlock != null) {
+          printRow(cellBlock);
+        }
+        cellBlock = new ArrayList<DiffCell>();
+      }
+      // Add this cell to the active cell-block.
+      cellBlock.add(diffCell);
+      col = diffCell.colIndex + 1;
```

Comment:
Oh so it's printing the diff per differing "contiguous cell block". That's not the impression I got from the README - I thought all the diff cells in one row are printed together. Might want to update the READE

Reply:
I'll try to make that clearer.  Yes, the idea is to group as many contiguous differing cells as possible in the same row together.  If I had a more powerful diff engine to play with, I'd try to expand that to blocks of rectangularly-contiguous cells (_e.g._, A1-C3 instead of A1-A3, B1-B3, and C1-C3).  But that's something for another day.

Reply:
README has been updated.

Reply:
Right 2d contiguous blocks - I could see some good uses for it - e.g. to simplify and reduce diff output. But in general the human brain may not powerful enough to process that output